OCC startup (istep 21.1)

SergiiDmytruk opened the pull request:
Includes #89.

SergiiDmytruk committed:
istep 16.2

Change-Id: I082ed14d5c6f7fa26a6edefd6d73f725b0b36cd9
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

Commit:
Add OCC operations

Signed-off-by: Igor Bagnucki <igor.bagnucki@3mdeb.com>

SergiiDmytruk committed:
Beginning of istep 21.1

Change-Id: Ia342e6befc92c52dcced360dbc5c658ddf31d6b8
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Continue with start_pm_complex()

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Finish with setup_memory_boot()

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Implement special_wakeup_all(enable=false)

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
pm_corequad_init()

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Implement pstate_gpe_init()

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Implement pm_pba_init()

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Initialize MCS bits in configuration vector

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Implement wait_for_occ_checkpoint()

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Implement poll_occ()

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
The bulk of sending commands to OCC

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Finish implementing communication with OCC

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Implement switching OCC to active state

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Initialize user power limit (no limit)

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Start implementing send_occ_config_data()

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Add PCAP config data

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Do not auto-start Pstate GPE protocol

Hostboot doesn't seem to do it and it times out anyway.

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Fix SCOM addrs in pba_slave_setup_runtime_phase()

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Fix bugs on constructing bootcode for OCC

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Simplify pba_slave_setup_runtime_phase()

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Fix stack usage errors

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Fix PPC_PLACE() macros in pb_pba_init()

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Simplify pm_pba_bar_config()

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Add pm_pss_init()

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
pstate_gpe_init() fixes

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Fix wait_for_occ_checkpoint()

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Comments and formatting

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Bug fixes

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Pass APSS, MEMCFG and SYSCFG data to OCC

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
The rest of configuration data

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Process OCC errors

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
More error reporting

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Improve get_gpu_msg_data()

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Set error masks like Hostboot does

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Better logging

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Restart SGPE before starting OCC

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Don't add istep 16.2 here

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Remove SGPE reset code

It didn't work and this can be done differently.

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Start OCC in istep 15.4

This makes resetting it later unnecessary.

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Some code improvements

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Remove debug output

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Add informational messages

It takes some time to start OCC.

Change-Id: I1609349c673b4f0295e4e743add61f66a66aa6b1
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Formatting corrections for occ.h/occ.c

Change-Id: I56651886bc752ce9f17707ff0393abfdb65cb51d
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
A couple of corrections

Change-Id: I4ab8a73cb73ccd486f08713920e8c389adff22b7
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

krystian-hebel commented on src/include/cpu/power/occ.h:
```diff
@@ -0,0 +1,86 @@
+#define OCC_405_SRAM_ADDRESS (0xFFF40000)
+#define OCC_OFFSET_MAIN_EP (0x6C)
+#define OCC_BRANCH_INSTR (0x4B00000200000000)
+#define BRANCH_ADDR_MASK (0x00FFFFFC)
```
These are wrong, should be 0x48000002 and 0x03FFFFFC respectively, this error was copied from Hostboot. Nonetheless, it is not actually used so can be removed.

krystian-hebel commented on src/include/cpu/power/occ.h:
```diff
@@ -0,0 +1,86 @@
+	EP03_CHIPLET_ID,
+	EP04_CHIPLET_ID,
+	EP05_CHIPLET_ID
+};
```
There are 12 EXs, which are accessed through quad chiplets. Names are misleading, as well as use of this table in clear_occ_special_wakeups() - see comment there.

krystian-hebel commented on src/include/cpu/power/occ.h:
```diff
@@ -0,0 +1,86 @@
+    const uint32_t address,
+    uint64_t * buffer,
+    size_t data_length);
+uint64_t makeStart405Instruction(void);
```
Unused, please remove.

krystian-hebel commented on src/include/cpu/power/occ.h:
```diff
@@ -0,0 +1,86 @@
+    PU_OCB_PIB_OCBCSR1_OR,
+    PU_OCB_PIB_OCBCSR2_OR,
+    PU_OCB_PIB_OCBCSR3_OR
+};
```
AFAICT only OCB0 is ever used, drop these tables and use appropriate register directly.

krystian-hebel replied:
Also, if any register address is not needed outside of occ.c, move its definition there.

SergiiDmytruk replied:
Moved more stuff to `occ.c`, which simplified things. I was just expecting #89 to be merged first and avoided modifying `occ.c` too much.

krystian-hebel replied:
We will have to divide the code even more for upstreaming and probably move the rest of OCC functions to `occ.c`, but for now this should be good enough.

krystian-hebel commented on src/soc/ibm/power9/Makefile.inc:
```diff
@@ -9,6 +9,7 @@ romstage-y += romstage.c
 romstage-y += mvpd.c
 romstage-y += vpd.c
 romstage-y += powerbus.c
+romstage-y += occ.c
```
Why is it needed for romstage?

SergiiDmytruk replied:
This is from #89, I removed this line in a later commit.

krystian-hebel replied:
Huh, I have no idea how but I managed to review older version, sorry about that.

krystian-hebel commented on src/soc/ibm/power9/occ.c:
```diff
@@ -0,0 +1,109 @@
+            EX_PPM_SPWKUP_OCC,
+            read_scom_for_chiplet(EX_CHIPLETS[chiplet_index],
+            EX_PPM_SPWKUP_OCC) & ~PPC_BIT(0));
+    }
```
As mentioned in the header, there are 12 EXs, which are accessed through even core chiplets. This loop should be something like this:

```
for (size_t chiplet_index = 0; chiplet_index < MAX_CORES /* 24 */; chiplet_index += 2) {
    write_scom_for_chiplet(EP00_CHIPLET_ID + chiplet_index, EX_PPM_SPWKUP_OCC, ....
}
```
or
```
for (size_t ex_index = 0; ex_index < MAX_EXS /* 12 */; ++ex_index) {
    write_scom_for_chiplet(EP00_CHIPLET_ID + ex_index * 2, EX_PPM_SPWKUP_OCC, ....
}
```

krystian-hebel commented on src/soc/ibm/power9/homer.c:
```diff
@@ -997,6 +1076,244 @@ static void istep_16_1(int this_core)
+	};
+
+	write_scom_direct(PBA_BARs[index], bar_addr & 0x1FFFFFFFFFFFFFFFull);
+	write_scom_direct(PBA_BARMSKs[index], 0x300000);
```
According to dumps, this is 8M for BAR2 (0x700000). This is also set [here](https://github.com/3mdeb/coreboot/blob/a11f8a632a089c3de818f467f6d6abc658222588/src/soc/ibm/power9/homer.c#L2128) - one of these should be enough.

krystian-hebel commented on src/soc/ibm/power9/homer.c:
```diff
@@ -997,6 +1076,244 @@ static void istep_16_1(int this_core)
+static void load_pm_complex(struct homer_st *homer)
+{
+	/* Common OCC area is located right after HOMER image */
+	uint8_t *common_area = (uint8_t *)homer + sizeof(*homer);
```
Hostboot reserves memory for 8 HOMER images (4M each) and one common area (8M) after the last one. For Talos we will need two HOMERs when second CPU is installed, so it shouldn't be hardcoded. Second CPU will also use different address of HOMER, but the same common area, so for now it will be enough to just pass pointer to common_area to load_pm_complex() and hardcode it at the highest possible level.

krystian-hebel commented on src/soc/ibm/power9/homer.c:
```diff
@@ -997,6 +1076,244 @@ static void istep_16_1(int this_core)
+static void check_proc_config(struct homer_st *homer)
+{
+	uint64_t vector_value = INIT_CONFIG_VALUE;
+	/* XXX: how come this points into padding area? */
```
This explains last unaccounted differences between our HOMER and one prepared by Hostboot somewhere after 15.2

krystian-hebel commented on src/soc/ibm/power9/homer.c:
```diff
@@ -997,6 +1076,244 @@ static void istep_16_1(int this_core)
+
+		write_scom(spwkup_addr, 0);
+		(void)read_scom(spwkup_addr);
+	}
```
Is it supposed to work on cores, and not EXs as clear_occ_special_wakeups() did? I don't see this in dumps at all.

The final read is some kind of workaround? If yes, please add a comment.

SergiiDmytruk replied:
I probably did iteration over cores because they are used for computing address in `src/import/chips/p9/procedures/hwp/pm/p9_cpu_special_wakeup_ex.C` (code related to wakeups is rather convoluted) and logs have this as if they iterate over cores as well:
```
553.73054|SCAN|TRACE : PUTSCOM     :  pu:k0:n0:s0:p00 : 00000000210F010B 0000000000000000
553.73056|SCAN|TRACE : GETSCOM     :  pu:k0:n0:s0:p00 : 00000000210F010B 0000000000000000

553.73084|SCAN|TRACE : PUTSCOM     :  pu:k0:n0:s0:p00 : 00000000230F010B 0000000000000000
553.73086|SCAN|TRACE : GETSCOM     :  pu:k0:n0:s0:p00 : 00000000230F010B 0000000000000000

553.73114|SCAN|TRACE : PUTSCOM     :  pu:k0:n0:s0:p00 : 00000000240F010B 0000000000000000
553.73116|SCAN|TRACE : GETSCOM     :  pu:k0:n0:s0:p00 : 00000000240F010B 0000000000000000

553.73145|SCAN|TRACE : PUTSCOM     :  pu:k0:n0:s0:p00 : 00000000270F010B 0000000000000000
553.73147|SCAN|TRACE : GETSCOM     :  pu:k0:n0:s0:p00 : 00000000270F010B 0000000000000000
```
Hostboot seems to send wakeup to functional core of an EX instead of the first one.

Final read adds a delay, copied comment from Hostboot.

krystian-hebel commented on src/soc/ibm/power9/homer.c:
```diff
@@ -997,6 +1076,244 @@ static void istep_16_1(int this_core)
+
+static void pm_occ_control_start_from_mem(void)
+{
+	write_scom(OCBCSRn_OR[0], PPC_BIT(OCB_PIB_OCBCSR0_OCB_STREAM_MODE));
```
Use `PU_OCB_PIB_OCBCSR0_OR` directly.

krystian-hebel commented on src/soc/ibm/power9/occ.c:
```diff
@@ -0,0 +1,109 @@
+{
+    write_scom(OCBCSRn_OR[0], PPC_BIT(OCB_PIB_OCBCSR0_OCB_STREAM_MODE));
+    write_scom(OCBCSRn_CLEAR[0], PPC_BIT(OCB_PIB_OCBCSR0_OCB_STREAM_TYPE));
+    write_scom(OCBARn[0], (uint64_t)i_ocb_bar << 32);
```
If this is always for OCB0, use register names directly.

krystian-hebel commented on src/soc/ibm/power9/occ.c:
```diff
@@ -0,0 +1,109 @@
+                break;
+            }
+            // Hostboot has delay of 0 here
+            wait_us(1, false);
```
Does it work without this delay? Maybe it was used for simulation only.

SergiiDmytruk replied:
Haven't tried without the delay, will test. I'm surprised it works at all... bits are checked with `& bit_pos` instead of `& PPC_BIT(bit_pos)`. Haven't looked at it closely before.

SergiiDmytruk replied:
Removed the delay. Seems to make no difference.

krystian-hebel commented on src/soc/ibm/power9/occ.c:
```diff
@@ -0,0 +1,109 @@
+    // the branch instruction must be '2' according to the OCC instruction
+    // set manual.
+    return OCC_BRANCH_INSTR | (((uint64_t)(BRANCH_ADDR_MASK & l_epAddr)) << 32);
+}
```
Unused, please remove.

krystian-hebel commented on src/soc/ibm/power9/homer.c:
```diff
@@ -997,6 +1076,244 @@ static void istep_16_1(int this_core)
+
+static uint32_t ppc_mtspr(uint16_t rs, uint16_t spr)
+{
+	enum { MTSPR_CONST1 = 467 };
```
Maybe use 0x7C0003A6 as MTSPR_OP and drop this? This is extended opcode, without it instruction isn't `mtspr` anyway.

krystian-hebel commented on src/soc/ibm/power9/homer.c:
```diff
@@ -997,6 +1076,244 @@ static void istep_16_1(int this_core)
+
+static uint32_t ppc_bctr(void)
+{
+	enum { BCCTR_CONST1 = 528 };
```
Same as above.

krystian-hebel commented on src/soc/ibm/power9/homer.c:
```diff
@@ -997,6 +1076,244 @@ static void istep_16_1(int this_core)
+static uint32_t ppc_b(uint32_t target_addr)
+{
+	uint32_t inst;
+	inst = BR_OP << (31 - 5);
```
BR_OP is already shifted, so:
```suggestion
	inst = BR_OP;
```
No idea why it works in this form...

SergiiDmytruk replied:
This is from an old commit.

SergiiDmytruk committed:
Remove makeStart405Instruction() and related

Change-Id: I632fbad29216554ba44055ce803d912a4246da4c
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Extract ops.h

Change-Id: Ia3d53035ac34e64cd5c224e1423526990f62df06
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Move some OCC-specific code from homer.c to occ.c

Change-Id: Iebbba0ecaa235c37c70d182ec03022d852402971
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Update OP codes

Change-Id: Idfe51a377fd01d945ba888f4cc613742f067a955
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Fixes for sending wakeups

Change-Id: I5a011866c2453801a5e998fe87550b3e288f970d
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Don't set HOMER BAR second time

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Remove extra delay in check_ocb_mode()

Change-Id: I53c1c46492d57032d02bf3fc514933e4f4f50597
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

krystian-hebel commented on src/soc/ibm/power9/homer.c:
```diff
@@ -997,6 +1090,1376 @@ static void istep_16_1(int this_core)
+		 * 18 - 19    : PCB interrupt
+		 * 20,22,24,26: InterPPM Ivrm/Aclk/Vdata/Dpll enable
+		 */
+		write_scom_for_chiplet(quad_chiplet, EQ_QPPM_QPMMR_CLEAR, 0xfffe3aa000000000ull);
```
Here and for all other SCOMs: I'd rather see PPC_BIT(MASK) macros used for this, as long as we know the layout of the register. That way it is easier to understand and check if code does what comment says without manually counting bit positions which is error prone, especially when value doesn't have all 16 hex-digits. Consecutive fields can be covered with one macro, e.g. PPC_BITMASK(0,14) in this case.

If that is not possible, please use capital letters for hex values.

krystian-hebel commented on src/soc/ibm/power9/homer.c:
```diff
@@ -997,6 +1090,1376 @@ static void istep_16_1(int this_core)
+			write_scom_for_chiplet(core_chiplet, C_CPPM_CSAR_CLEAR, 0x1b00000000);
+
+			/* Restore CORE PPM Error Mask */
+			err_mask = 0xfff00000; // from Hostboot's log
```
```suggestion
			err_mask = 0xFFF00000; // from Hostboot's log
```

krystian-hebel commented on src/soc/ibm/power9/homer.c:
```diff
@@ -997,6 +1090,1376 @@ static void istep_16_1(int this_core)
+	write_scom(PU_GPE2_PPE_XIXCR, PPC_PLACE(RESUME, 1, 3));
+
+	/* Does this need to be such a long time? */
+	wait_ms(20*TIMEOUT_COUNT,
```
```suggestion
	wait_ms(PGPE_POLLTIME_MS * TIMEOUT_COUNT,
```

I wonder if we should add a version of wait_ms() to coreboot that checks condition periodically instead of using busy loop. For memory training busy reading significantly slowed down the training. The issue is that we can't use delay() in ramstage, at least not in the same form. Still, this is not part of this task, I'm just thinking out loud :slightly_smiling_face:

krystian-hebel commented on src/soc/ibm/power9/homer.c:
```diff
@@ -997,6 +1090,1376 @@ static void istep_16_1(int this_core)
+	data |= PPC_PLACE(PBA_BCE_OCI_TRANSACTION_64_BYTES, 21, 2); // bcde_ocitrans
+	data |= PPC_PLACE(PBA_BCE_OCI_TRANSACTION_64_BYTES, 23, 2); // bcue_ocitrans
+	data |= PPC_BIT(8);                                         // en_marker_ack
+	data |= PPC_PLACE(PBA_OCI_MARKER_BASE >> 16, 18, 3);        // oci_marker_space
```
Is this correct? PBA_OCI_MARKER_BASE is 0x40070000, it won't fit in just 3 bits. Maybe just set it to 0x7 and skip shifting.

SergiiDmytruk replied:
That's how Hostboot does it. I guess `0x40070000` is a final address and `oci_marker_space` is an offset in 64 KiB units from a base address. But yes, it looks too complicated.

krystian-hebel commented on src/soc/ibm/power9/homer.c:
```diff
@@ -997,6 +1090,1376 @@ static void istep_16_1(int this_core)
+	}
+}
+
+static void pba_slave_setup_runtime_phase(void)
```
This is almost identical to the last lines of pba_reset(), except write gather is disabled for slave 0 there. As SGPE isn't started before pba_slave_setup_runtime_phase() anyway, could you remove lines 790-804? You did much better work describing those registers than me.

krystian-hebel replied:
Also, as these are not documented, please copy the description from comments below to [undocumented_registers.md](https://github.com/3mdeb/openpower-coreboot-docs/blob/main/devnotes/undocumented_registers.md)

SergiiDmytruk replied:
Ended up calling this function once on line 790. Calling it only in `pm_pba_init()` caused hang somewhere in `start_pm_complex()`.

krystian-hebel commented on src/soc/ibm/power9/homer.c:
```diff
@@ -997,6 +1090,1376 @@ static void istep_16_1(int this_core)
+		if ((read_scom_for_chiplet(nest, 0x0501080A) & PPC_BIT(0)) ||
+		    (read_scom_for_chiplet(nest, 0x0501080C) & PPC_BIT(0))) {
+			uint8_t pos = MCS_POS + mcs_i;
+			*conf_vector |= (0x8000000000000000ull >> pos);
```
```suggestion
			*conf_vector |= PPC_BIT(pos);
```
Here and below

krystian-hebel commented on src/soc/ibm/power9/homer.c:
```diff
@@ -997,6 +1090,1376 @@ static void istep_16_1(int this_core)
+	scom_and_or(PU_SPIPSS_P2S_CTRL_REG0, ~PPC_BITMASK(0, 11), PPC_BIT(2));
+	scom_and_or(PU_SPIPSS_P2S_CTRL_REG0 + 1, ~PPC_BITMASK(1, 3),
+		    PPC_BIT(0) | PPC_BIT(10) | PPC_BIT(12) | PPC_BIT(17));
+	scom_and(PU_SPIPSS_P2S_CTRL_REG0 + 2, ~PPC_BITMASK(0, 16));
```
Please add comment about which bits are set. Seems that some of them should be set with PPC_PLACE/PPC_SHIFT instead.

krystian-hebel commented on src/soc/ibm/power9/homer.c:
```diff
@@ -997,6 +1090,1376 @@ static void istep_16_1(int this_core)
+	}
+
+	/* Statuses of 0xE0-EF are reserved for OCC exceptions */
+	if ((status & 0xF0) == 0xE0)
```
Is it even possible? parse_occ_response() checks for OCC_RC_SUCCESS and returns false, so previous `if` will be taken in that case.

krystian-hebel commented on src/soc/ibm/power9/occ.c:
```diff
@@ -0,0 +1,238 @@
+	enum { PPM_SPWKUP_FSP = 0x200F010B };
+
+	for (int i = 0; i < MAX_CORES_PER_CHIP; ++i) {
+		uint32_t spwkup_addr = PPM_SPWKUP_FSP + 0x01000000 * i;
```
This line is useless, `*_for_chiplet()` functions already add proper offset. Remove it and use `PPM_SPWKUP_FSP` instead of `spwkup_addr`

krystian-hebel commented on src/soc/ibm/power9/occ.c:
```diff
@@ -0,0 +1,238 @@
+	return inst;
+}
+
+static uint32_t ppc_ori(uint16_t rs, uint16_t ra, uint16_t data)
+{
+	uint32_t inst;
+	inst = ORI_OP;
+	inst |= rs << (31 - 10);
+	inst |= ra << (31 - 15);
+	inst |= data;
+	return inst;
+}
+
+static uint32_t ppc_mtspr(uint16_t rs, uint16_t spr)
+{
+	uint32_t temp = ((spr & 0x03FF) << (31 - 20));
+
+	uint32_t inst;
+	inst = MTSPR_OP;
+	inst |= rs << (31 - 10);
+	inst |= (temp & 0x0000F800) << 5;  // Perform swizzle
+	inst |= (temp & 0x001F0000) >> 5;  // Perform swizzle
+	return inst;
+}
+
+static uint32_t ppc_bctr(void)
+{
+	uint32_t inst;
+	inst = BCCTR_OP;
+	inst |= 20 << (31 - 10); // BO
+	return inst;
+}
+
+static uint32_t ppc_b(uint32_t target_addr)
+{
+	uint32_t inst;
+	inst = BR_OP;
+	inst |= (target_addr & 0x03FFFFFF);
+	return inst;
+}
```
Please make those inline and move to ops.h. I can think of at least one use for ppc_mtspr elsewhere.

SergiiDmytruk committed:
Move ppc_* functions from occ.c to ops.h

Change-Id: Ic6ebdd8e6fa2f8347f60d6a2435ce7c4ef1eb19d
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Simplify special_occ_wakeup_disable()

Change-Id: I93427b99d18f3ca8b35c5cc9d0414a8d868f7f6d
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Comments and corrections for pm_pss_init()

Change-Id: I75a7bbf1474ffe556356f04e28d23bd47d14004f
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Remove duplicated pba slave setup

Change-Id: I8a634e023dd8cecfa771c68ebe5134c1d4acf04e
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Fix deadcode in write_occ_cmd()

Change-Id: Ia13098cedf6e68889a4152dfb616710aebb488d1
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Various review corrections

Change-Id: Ia70f487001b965648ec779353149f2c62f996670
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Compute SCOM values using PPC_* macros

Change-Id: I306de2b11cd98cc99c3262dc74dabcf8c22565d7
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Use correct PB freq in one more place

Change-Id: I212fec18384c81488b5f3de1631dda1df08248d0
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Initialize PBA slaves earlier (on PBA reset)

Change-Id: Ia0e991efe26b8abf6f020cb10eb82f951674e25b
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Add pm_occ_fir_init()

Change-Id: I483b9ce660a7d840a52437f1de0428df2ea84706
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Add pm_pba_fir_init()

Also fix a bug in pm_occ_fir_init() and move the two functions to a
separate file.

Change-Id: Idbbbe002ed673beaad18ccd144f193a290baf969
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Make SCOM writes match Hostboot

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

krystian-hebel commented on src/soc/ibm/power9/fir.c:
```diff
@@ -0,0 +1,195 @@
+/* SPDX-License-Identifier: GPL-2.0-only */
```
Almost each component has it's own set of FIR, having this in separate file is misleading. Please move it to occ.c. Same goes for fir.h.

krystian-hebel reviewed (changes requested):
One small change, otherwise LGTM

SergiiDmytruk committed:
Merge fir.c into occ.c

Change-Id: I3e60bd4b2299691cb18df98c218514671eb3a440
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

krystian-hebel commented:
Let's wait for CI tests before merging.
